### ISSUE TITLE
Issues with custom dataset

### ISSUE THREAD

**Issue by Xtaltec** [gbouritsas/Neural3DMM#5]
Hi @gbouritsas. First of all, thank you for your great work.

I was trying to run the code with a custom dataset of medical meshes but the reconstructions I get during testing are not very good, at least when checking visually. My mesh dataset is completely registered, sharing both amount of nodes and connectivity, but it is not watertight. Could that be an issue?

![image](https://user-images.githubusercontent.com/42573295/112370822-249f9a80-8cde-11eb-9894-d5c22706781e.png)

**Comment by gbouritsas** [comment]
Hi @Xtaltec, thanks for your interest in our work. 

I am trying to understand the attached image. I probably got something wrong here, but why does your GT have two disconnected components and the prediction has only one? That cannot happen if all the meshes share the same topology, right? Could you explain what is each one of the 3 meshes that you attached? I don't think that having non-watertight meshes is an issue. We actually experimented with faces that are open surfaces and faces with disconnected components. I can't guarantee that this is not a problem, but I doubt it. 

How does your training error behave? Is it also high, or is it just the test set that suffers from bad predictions? Do your train and test sets come from the same distribution? Does a linear baseline (e.g. PCA) generalise well to test data? Btw, what kind of surfaces are you modelling?

**Comment by Xtaltec** [comment]
Thanks for your prompt response! I apologize in advance if the first post was not clear enough. 

I am working with meshes from the left atrial appendage, a heterogeneous tubular structure within the left atrium. In the previous image, the two meshes in the left and the center correspond to the same ground truth case. I added two distinct views aiming to highlight that the mesh is open at one of the ends.

![image](https://user-images.githubusercontent.com/42573295/112401234-61818680-8d0a-11eb-826e-8c06f3039774.png)

The cohort I'm working with at the moment consists of only 250 cases. I have already experimented with PCA models and although I have to check the reconstruction error, I got acceptable results.

On the other hand, I attach the training-testing results below. I gave a look at the code but I'm not sure if the normalized L1 loss from testing is somewhat different from the one in training?

![image](https://user-images.githubusercontent.com/42573295/112399844-537e3680-8d07-11eb-93f7-0162312d211d.png)

**Comment by sergeybok** [comment]
Are you predicting normalized outputs or predicting the vertices' locations themselves? In our code it should grab the mean and std-dev of the distribution and be deforming the mean so unless you changed it that should be fine but something to double check because predicting the normalized outputs and multiplying by the std-dev and adding the mean is much easier to learn than the XYZ locations themselves. 

Secondly, your dataset is much smaller than ours was, what's your 250 train / val / test split? It is very likely overfitting the data based on what you posted, so you might need to try more aggressive regularization. Have you also visualized one of the samples from the training set, does it also not look good? If it doesn't then there's probably something wrong, if it looks good but the test is bad, then it's overfitting and either use a smaller configuration of the model, higher weight decay, or try to get more data (which I understand might not be viable but pretty standard response for deep learning methods).

**Comment by Xtaltec** [comment]
I have checked the training results and they are also distorted with respect to the original mesh. I have only slightly modified the original code to fit my dataset but have not touched anything in terms of standardizing the data. 

I was thinking that the problem might have been caused by saving the vertex sets in train.npy and test.npy, since I am using the pyvista library with which I am more familiar. But I have delved into the downsampling code and the order of the vertices and edges seems exactly the same as with mpi-mesh.

I also checked if the downsampled meshes were ok, I attach the corresponding image below. I then tried running it with the downsampling reduced by half in case it was too aggressive, but the results are similar. 

![image](https://user-images.githubusercontent.com/42573295/112679047-036cb480-8e6c-11eb-91ea-8c2b96e4bd02.png)

**Comment by gbouritsas** [comment]
Ok, so as far as I understand you are also having trouble fitting your training data as well. Some more suggestions:

- What's the size of your validation set? The default is 100, which, compared to 250, is quite large, so you would be wasting too much data for validation.
- Have you tried modifying the network or the optimisation parameters? You can try to reduce the widths of the convolutional layers. Or maybe try to increase the size of the latent space (I can see that now the default is 16). Another thing that might be helpful when debugging is to always compare with the PCA error when using the same latent space size.

Btw, make sure that you set args['normalization'] = True as @sergeybok mentioned.

**Comment by Xtaltec** [comment]
Hi!

Apologies for taking so long to reply. I managed to solve the issue. I wasn't multiplying by the standard deviation and adding the mean to the test predictions, I assumed the code was doing it for me. 

Thanks again for providing the code and I'm sorry for the hassle!
